### DISCUSSION TITLE
VS2022 vcpkg integration?

### DISCUSSION THREAD

**Discussion by nurbles** [microsoft/vcpkg#21337] [Q&A]
I've just set up a new, clean Windows 10 system and installed Visual Studio 2022 Community to start testing it with our solutions. We use a few packages from vcpkg, so I followed the instructions from [github/vcpkg](https://github.com/microsoft/vcpkg) to install vcpkg and the few x64-windows packages we've been using in VS2019 (and VS2017 before that.) After installing the packages, I tried the `vcpkg integrate install` command and get this:

`Error: failed to copy targets file to C:\Program Files (x86)\MSBuild/Microsoft.Cpp/v4.0/V140/ImportBefore/Default/vcpkg.system.props`

Since VS2022 is now 64 bit, it should not be located in `Program Files (x86)` (and it is not.)  The closest path I could find to what vcpkg is apparently hard-coded to look for was `C:\Program Files\Microsoft Visual Studio\2022\Community\Msbuild\Current\Imports\Microsoft.Common.Props\ImportBefore` although it appears that the entire folder path that vcpkg wants has also been created (and is completely empty) as well. Since the target folder in the error message exists, and the command is running with admin rights, I do not understand why it would fail to copy its file.

So, what have I done wrong; do I not understand; or are we simply still waiting for a VS2022-compatible version of vcpkg?  Thanks!

**Comment by tecandrew** [comment]
try setting the VCPKG_VISUAL_STUDIO_PATH` env variable, pointing it to 
```
C:\Program Files\Microsoft Visual Studio\2022\Community
```

**Reply by nurbles** [reply]
I tried that setting but it appears to have had no effect on the results.  If interested, you can see what I entered and the error in [this image](https://www.svalli.com/files/vcpkg-error.png).  Thanks for trying, though!

FWIW, I've done this successfully using the same vcpkg location (`M:\vcpkg`) with both VS2017 and VS2019, but they were both 32 bit applications located where vcpkg seems to be expecting to find them.  I didn't see a special vcpkg64 for VS2022 -- is that a thing that I should have searched for?

**Reply by tecandrew** [reply]
ahh, dang. thought that would help. currently running into issues with vcpkg on a fresh container of win10 + visual studio, but not vs 2022. this might be the case of being a bit too cutting-edge 😅

**Reply by nurbles** [reply]
#a bit too cutting edge
Yup, that's why I've built a test system. I guess I'm just surprised that none of the vcpkg developers tested integration with VS2022 while it was in preview. After all, both are from Microsoft, so as a developer I kinda expect them to keep their own tools in sync.

But, since VS2022 is still contains many things that haven't evolved much (or at all) in something like 20+ years (I'm looking at you, Resource Editor and Fonts&Colors settings) I guess I'm just a dreamer. I see that as a good thing -- after over 40 years as a professional developer, I still have hope that others will fix the old, unglamorous things.  :-) 

Again, thanks for trying!

**Comment by ras0219-msft** [comment]
The failure to copy should only affect VS2015 (that's what the v140 in the path indicates); VS2017, VS2019, and VS2022 all have built-in support for vcpkg that should work even if that particular file copy fails.

Have you tried using the libraries in the IDE after receiving the error?

**Reply by nurbles** [reply]
All of my projects already had the vcpkg include and library paths manually entered into the VC Directories fields for the project settings, so they built fine. However, the 'vcpkg' panel that had appeared in the project settings under VS2019 does not exist.

I guess I was really reporting this because following the instructions for vcpkg install and integration on a brand new, sparkling clean install of VS2022 (and the Windows 10 underneath it, too!) failed. That seems undesirable.

It is entirely likely that I did not install some VS2022 component that is required (but not mentioned in the vcpkg instructions) as I only install the C/C++ desktop development 'package' within the VS Installer.  If there is an assumption of some other package being installed, it should probably at least be mentioned in the vcpkg instructions.

If vcpkg has no requirements beyond the base Visual Studio product for the 'vcpkg integrate' to work, then I've encountered some other problem and I'm willing to try to help.  If there's any information I can collect from my system, just ask.

**Reply by ras0219-msft** [reply]
Could you run `vcpkg integrate install --debug` and include the full output with usernames and sensitive data redacted?

You're absolutely right that we want the initial instructions to work with the environment you've described. The C++ workload and English language pack should be the only required components.

**Reply by nurbles** [reply]
Here you go, I hope it helps.  Oh, if it isn't clear from the output below, I use M:\vcpkg as the root of all of my vcpkg things.
```
M:\vcpkg>vcpkg integrate install --debug
[DEBUG] Feature flag 'binarycaching' unset
[DEBUG] Feature flag 'manifests' unset
[DEBUG] Feature flag 'compilertracking' unset
[DEBUG] Feature flag 'registries' unset
[DEBUG] Feature flag 'versions' unset
[DEBUG] BuiltinRegistry initialized with: ""
[DEBUG] Using vcpkg-root: M:\vcpkg
[DEBUG] Using installed-root: M:\vcpkg\installed
[DEBUG] BuiltinRegistry initialized with: ""
[DEBUG] Using buildtrees-root: M:\vcpkg\buildtrees
[DEBUG] Using downloads-root: M:\vcpkg\downloads
[DEBUG] Using packages-root: M:\vcpkg\packages
[DEBUG] Using scripts-root: M:\vcpkg\scripts
[DEBUG] Using ports-root: M:\vcpkg\ports
[DEBUG] Using versions-root: M:\vcpkg\versions
[DEBUG] Failed to open: C:\Program Files (x86)\MSBuild/Microsoft.Cpp/v4.0/V140/ImportBefore/Default/vcpkg.system.props
Error: failed to copy targets file to C:\Program Files (x86)\MSBuild/Microsoft.Cpp/v4.0/V140/ImportBefore/Default/vcpkg.system.props
[DEBUG] D:\a\_work\1\s\src\vcpkg\commands.integrate.cpp(233)
[DEBUG] Exiting after %s (%d us)


M:\vcpkg>
```
This system had Windows 10 Pro installed on a freshly formatted drive (and updated to current) and then Visual Studio 2022 was installed.  Finally, I downloaded vcpkg and started following instructions.  Let me know if there's anything else I can get, try or otherwise do to help.  Thanks!

**Reply by RobertBernstein** [reply]
**Update**: Never mind.  I re-ran `vcpkg integrate install` from an elevated command prompt and now it works fine.

Any update on this?  I am running VS 2022 Preview and it's not finding my VcPkg installation at all.  I ran `vcpkg integrate install --debug`, but nothing looked obviously wrong.

```
C:\Dev\vcpkg>vcpkg integrate install --debug
[DEBUG] Feature flag 'binarycaching' unset
[DEBUG] Feature flag 'manifests' unset
[DEBUG] Feature flag 'compilertracking' unset
[DEBUG] Feature flag 'registries' unset
[DEBUG] Feature flag 'versions' unset
[DEBUG] Using vcpkg-root: C:\Dev\vcpkg
[DEBUG] Failed to open: C:\Dev\vcpkg\vcpkg-bundle.json
[DEBUG] Bundle config: readonly=0, usegitregistry=0, embeddedsha=nullopt
[DEBUG] Using scripts-root: C:\Dev\vcpkg\scripts
[DEBUG] Using downloads-root: C:\Dev\vcpkg\downloads
[DEBUG] Using scripts-root: C:\Dev\vcpkg\scripts
[DEBUG] Using builtin-ports: C:\Dev\vcpkg\ports
[DEBUG] Using builtin-registry: C:\Dev\vcpkg\versions
[DEBUG] Using installed-root: C:\Dev\vcpkg\installed
[DEBUG] Using packages-root: C:\Dev\vcpkg\packages
[DEBUG] Using buildtrees-root: C:\Dev\vcpkg\buildtrees
Applied user-wide integration for this vcpkg root.

All MSBuild C++ projects can now #include any installed libraries.
Linking will be handled automatically.
Installing new libraries will make them instantly available.

CMake projects should use: "-DCMAKE_TOOLCHAIN_FILE=C:/Dev/vcpkg/scripts/buildsystems/vcpkg.cmake"
[DEBUG] D:\a\_work\1\s\src\vcpkg\commands.integrate.cpp(305)
[DEBUG] Time in subprocesses: 0 us
[DEBUG] Time in parsing JSON: 0 us
[DEBUG] Time in JSON reader: 0 us
[DEBUG] Time in filesystem: 37513 us
[DEBUG] Time in loading ports: 0 us
[DEBUG] Exiting after 154.8 ms (127034 us)
```

**Reply by jeroen-z** [reply]
Think I stumbled upon the same issue. Roughly it's like this, I guess: vcpkg default to x86 triplet, whereas VS2022 (also Preview) default to x64 build. When I installed required packages with option `---triplet=x64-windows` I could get things going again but YMMV.
